ext/standard/assert.c

```diff
@@ -171,41 +172,35 @@ PHP_FUNCTION(assert)
 
 		ZVAL_FALSE(&retval);
 
-		/* XXX do we want to check for error here? */
-		if (!description) {
+		if (!description_str && !description_obj) {
 			call_user_function(NULL, NULL, &ASSERTG(callback), &retval, 3, args);
 			zval_ptr_dtor(&(args[2]));
 			zval_ptr_dtor(&(args[0]));
-		} else {
-			ZVAL_STR(&args[3], zval_get_string(description));
+		} else if (description_str) {
+			ZVAL_STR(&args[3], description_str);
 			call_user_function(NULL, NULL, &ASSERTG(callback), &retval, 4, args);
 			zval_ptr_dtor(&(args[3]));
 			zval_ptr_dtor(&(args[2]));
 			zval_ptr_dtor(&(args[0]));
+		} else {
+			zend_argument_type_error(2, "must be of type string for the ASSERT_CALLBACK mode");
```

Comment:
This error and the same for ASSERT_WARNING seems pretty problematic. Whether an exception is passed is controlled by the assert() caller, while the assertion mode is controlled by a global setting. I don't think we can error in such a case.

Reply:
Another possibility would be to always throw the exception if it has been specified, without honoring warning/callback.

Reply:
I'm not sure which one I'd prefer. Although, since we're moving away from warnings anyway... maybe your version makes a little bit more sense. + exception messages can be empty easily, so maybe it's better not to go down in that rabbit hole.

Reply:
I went with your suggestion. I'm not sure I can `RETURN_THROWS()` after throwing the exception, but I'll come back to the question a bit later :)